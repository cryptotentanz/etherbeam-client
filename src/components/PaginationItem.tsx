import React, { FC } from 'react'
import {
  faAngleDoubleLeft,
  faAngleDoubleRight,
  faAngleLeft,
  faAngleRight,
  faEllipsisH,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons'
import ListItem from './ListItem'
import Button from './Button'

type PageProp = 'first' | 'previous' | 'ellipsis' | 'next' | 'last' | number

interface PaginationItemProps {
  page: PageProp
  action?: () => void
  selected?: boolean
  disabled?: boolean
}

const PaginationItem: FC<PaginationItemProps> = ({ page, action, selected, disabled }: PaginationItemProps) => {
  const getProps = (): { description: string | null; icon: IconDefinition | null } => {
    switch (page) {
      case 'first':
        return { description: 'First page', icon: faAngleDoubleLeft }
      case 'previous':
        return { description: 'Previous page', icon: faAngleLeft }
      case 'ellipsis':
        return { description: null, icon: faEllipsisH }
      case 'next':
        return { description: 'Next page', icon: faAngleRight }
      case 'last':
        return { description: 'Last page', icon: faAngleDoubleRight }
      default:
        return {
          description: `Page ${page}`,
          icon: null,
        }
    }
  }

  const { description, icon } = getProps()

  return (
    <ListItem>
      <Button
        link
        color="secondary"
        description={description}
        action={action}
        label={typeof page === 'number' ? page.toString() : null}
        icon={typeof page !== 'number' ? icon : null}
        disabled={disabled}
        active={selected}
      />
    </ListItem>
  )
}

export default PaginationItem
