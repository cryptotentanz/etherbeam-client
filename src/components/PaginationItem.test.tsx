import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import PaginationItem from './PaginationItem'

describe('<PaginationItem />', () => {
  const onClick = jest.fn()

  let component: ShallowWrapper

  describe('Default', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page={5} action={onClick} />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('description')).toBe('Page 5')
      expect(button.prop('action')).toBe(onClick)
      expect(button.prop('disabled')).toBeFalsy()
      expect(button.prop('active')).toBeFalsy()
      expect(button.prop('label')).toBe('5')
    })

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('First', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="first" action={onClick} />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('description')).toBe('First page')
      expect(button.prop('icon')).toBeDefined()
    })
  })

  describe('Previous', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="previous" action={onClick} />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('description')).toBe('Previous page')
      expect(button.prop('icon')).toBeDefined()
    })
  })

  describe('Ellipsis', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="ellipsis" action={onClick} />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('description')).toBeNull()
      expect(button.prop('icon')).toBeDefined()
    })
  })

  describe('Next', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="next" action={onClick} />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('description')).toBe('Next page')
      expect(button.prop('icon')).toBeDefined()
    })
  })

  describe('Last', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page="last" action={onClick} />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('description')).toBe('Last page')
      expect(button.prop('icon')).toBeDefined()
    })
  })

  describe('Selected', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page={5} action={onClick} selected />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('disabled')).toBeFalsy()
      expect(button.prop('active')).toBeTruthy()
    })
  })

  describe('Disabled', () => {
    beforeAll(() => {
      component = shallow(<PaginationItem page={5} action={onClick} disabled />)
    })

    it('should render', () => {
      const button = component.find('Button')
      expect(button.prop('disabled')).toBeTruthy()
      expect(button.prop('active')).toBeFalsy()
    })
  })
})
