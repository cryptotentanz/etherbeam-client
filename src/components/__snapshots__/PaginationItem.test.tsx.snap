// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationItem /> Default should match snapshot 1`] = `
<li
  class="my-listitem "
>
  <button
    class="my-button-link my-secondary-fg my-clickable"
    title="Page 5"
  >
    <span
      class=""
    >
      5
    </span>
  </button>
</li>
`;
